Minimal speculative decoding implementation

daviswer opened the pull request:
Add support for base speculative decoding. Currently does not support ~~batching or~~ efficient kv-cache memory management. 
(12/8/23: batching has been added)

Demo is a notebook running Llama-7bF with speculator on cpu, using model weights that I have stored locally. Let me know how/if this should be changed to a demo format more proper for the repo

ani300 commented on fms/models/llama.py:
```diff
@@ -324,11 +325,15 @@ def forward(
+            out.append(output)
+        if len(out)==1:
+            return out[0]
+        return out
```
if we're changing the outputs of a model, we should either have all models doing the same, (aka gpt bigcode, roberta) or at least document the difference somewhere

daviswer replied:
True... I need some way to access the embeddings of the model and not just the logits. Whatever way is most appropriate here, I can then propagate to other (generative) models

nairbv replied:
IIUC this is a common use-case we plan to handle via a "headless" model instance rather than an extra parameter. I think this was required in HF and motivated the separate `_helper` function. @JRosenkranz probably has a view on how best to structure that (across all of our models)

ani300 commented on fms/modules/attention.py:
```diff
@@ -240,7 +250,7 @@ def forward(
         # if use_cache=True, we return the hidden_state as well as the kv cache
         if use_cache:
-            return out, (keys, values)
+            return out, (keys, values) if not fat_cache else (past_key_value_state[0], past_key_value_state[1], new_keys, new_values)
```
given that black will already split this line into a few, might as well type down the whole if-else

ani300 commented on fms/modules/speculator.py:
```diff
@@ -0,0 +1,89 @@
+        self.vsize = vocab_size
+        self.emb = nn.ModuleList([nn.Embedding(vocab_size, emb_dim) for _ in range(n_heads)])
+        self.proj = nn.ModuleList([nn.Linear(emb_dim * 2, emb_dim, bias=False) for _ in range(n_heads)])
+        self.head = nn.ModuleList([nn.Linear(emb_dim, vocab_size, bias=False) for _ in range(n_heads)])
```
out of curiosity, how did you pick the architecture for this?

daviswer replied:
Simplest and flattest recurrent MLP architecture I could come up with!

ani300 commented on fms/modules/speculator.py:
```diff
@@ -0,0 +1,89 @@
+            z = torch.cat([state, z], dim=2) # b k 2d
+            state = self.a(self.ln[i](self.proj[i](z))) # b k d
+            probs = F.log_softmax(self.head[i](state), dim=2) # b k v
+            probs, preds = probs.topk(topk[i], dim=2) # b k k'
```
do you need softmax only to get the top k tokens? don't the logits still conserve the a > b if softmax(a/{a,b}) > softmax(b/{a,b})?

ani300 commented on fms/modules/speculator.py:
```diff
@@ -0,0 +1,89 @@
+        ----
+        Generate tree of candidate sequences given latest base model embedding (state) and chosen token (ind).
+        Topk indicates # of tree "branches" at each head. 
+        k pares down the candidate list from prod(topk) to the top k most confident. 
```
maybe rename the second k to another letter, so as to not have the top-k for each head, and then the k best for the total combination?

ani300 commented on fms/modules/speculator.py:
```diff
@@ -0,0 +1,89 @@
+        # state: b n d
+        # inds: b n+2 (..., pred token, n+2, n+3)
+        out = []
+        for i in range(self.nheads):
```
can this be written as a fused operation? given that the inputs are the same for each head

daviswer replied:
I don't think so - the state vector gets updated with each iteration

ani300 commented on fms/utils/generation.py:
```diff
@@ -126,3 +126,205 @@ def truncate_after_eos(result, eos_token_id):
+    kwargs["past_key_value_states"] = past_key_value_states
+
+    bsize = input_ids.size(0)
+    n_gen = torch.zeros(bsize, device=input_ids.device).int()
```
you can pass the dtype to zeros() to save one copy

ani300 reviewed (commented):
I need to walk myself through the math on utils/generation.py to convince myself it's all fine, but the code looks good and I assume it works as expected. That being said, on top of the notebook, a script that mirrors `scripts/inference.py` but with speculation would be appreciated for the repo.

daviswer commented:
Added a `speculative_inference.py` demo script, and updated `speculative_generate()` to accept inputs of varying lengths. Note that this does make the setup for the prefill step slightly ridiculous, and mypy is complaining due to not linking the paths of a couple if/else trees. Will try to clean that up later

daviswer commented:
OK, tests should be passing now. Outstanding issue at this point is the question of broader support for `include_embeds` in generative models (beyond just Llama)

nairbv commented on fms/utils/generation.py:
```diff
@@ -126,3 +128,234 @@ def truncate_after_eos(result, eos_token_id):
     return result
+
+
+def speculative_generate(
```
seems very long, maybe could be broken up with some helper functions to make it easier to read (and could document sub-parts). e.g. maybe functions to get the mask, to get the position IDs, one for the loop that manipulates the kv cache.

nairbv commented on fms/utils/generation.py:
```diff
@@ -126,3 +128,234 @@ def truncate_after_eos(result, eos_token_id):
+
+def speculative_generate(
+    model: Union[Callable, torch.nn.Module],
+    input_ids: List[torch.Tensor],
```
could this be a tensor with a batch dimension instead of list of tensors?

nairbv commented on fms/utils/generation.py:
```diff
@@ -126,3 +128,234 @@ def truncate_after_eos(result, eos_token_id):
+        max_seq_len: the sequence length of the base model
+        new_tokens: number of tokens to generate
+        top_k: only score the top k candidates from the speculator
+        threshes: use top k predictions from each head to generate speculator candidate pool
```
this isn't very clear, at least to me

nairbv commented on fms/utils/generation.py:
```diff
@@ -126,3 +128,234 @@ def truncate_after_eos(result, eos_token_id):
+    new_tokens: int = 256,
+    top_k: int = 5,
+    threshes=[5, 3, 2],
+    verbose_dict=None,
```
this should probably have a more complete type annotation so it's clearer how to use it (e.g. Optional[Mapping[ ??, ??]]` ?)

nairbv commented on speculative-decode-demo-7b-chat.ipynb:
this should probably go in the notebooks/ directory

nairbv commented:
@daviswer since we're putting a speculator in fms-extras, should this PR be closed?
